src/lib/reasoners/theory.ml

```diff
@@ -369,31 +373,118 @@ module Main_Default : S = struct
          (* A new explanation in order to track the choice *)
          (c, size, CPos exp, ex_c_exp)) l
 
+  let register_optimized_split objectives u =
+    try
+      let x = Util.MI.find u.Th_util.order objectives in
+      assert (E.equal x.Th_util.e u.Th_util.e); (* and its Value ... *)
+      Util.MI.add u.Th_util.order u objectives
+    with Not_found ->
+      assert false
+
+
+  exception Found of Th_util.optimized_split
+
+  (* TODO: this function could be optimized if "objectives" structure
+     is coded differently *)
+  let next_optimization env ~for_model =
+    try
+      Util.MI.iter (fun _ x ->
+          match x.Th_util.value with
+          | Value _ ->
+            ()
+          | Pinfinity | Minfinity ->
+            (* We should block case-split at infinite values.
+               Otherwise, we may have soundness issues. We
+               may think an objective is unbounded, but some late
+               constraints may make it bounded.
+
+               An alternative is to only allow further splits when we
+               know that no extra assumptions will be propagated to
+               the env. Hence the test 'if for_model' *)
+            if for_model then ()
+            else
+              raise (Found { x with Th_util.value = Unknown })
+
+          | Unknown ->
+            raise (Found { x with Th_util.value = Unknown })
+        ) env.objectives;
+      None
+    with Found x ->
+      Some x
+
+  (* TODO: this function could be optimized if "objectives" structure
+     is coded differently *)
+  let partial_objectives_reset objectives to_flip =
+    match to_flip with
+    | None -> objectives
+    | Some {Th_util.opt_ord; _} ->
+      Util.MI.fold
+        (fun ord v acc ->
+           if ord < opt_ord then
+             (*don't change older optims that are still valid splits*)
+             acc
+           else
+             match v.Th_util.value with
+             | Th_util.Unknown -> acc (* not optimized yet *)
+             | Value _ -> Util.MI.add ord {v with value = Unknown} acc
+             | Pinfinity | Minfinity -> assert false (* may happen? *)
+        ) objectives objectives
 
   let look_for_sat ?(bad_last=None) ch env l ~for_model =
-    let rec aux ch bad_last dl env li =
+    let rec aux ~optimize ch bad_last dl env li =
```

Comment:
This additional boolean flags make an already convoluted function even more convoluted.  I would suggest removing the flag and defining (either local or mutually recursive with aux) functions `optimizing_split` and  `case_split`, and to do something like:

```ocaml
match next_optimization ~for_model env with
| Some opt_split -> optimizing_split ch dl env opt_split
| None -> case_split ch dl env
```

in `aux` and:

```ocaml
match CC.optimizing_split env.gamma_finite opt_split with
| Some x -> (* same as currently *)
| None -> if for_model then case_split ch dl env else { env with choices = List.rev dl }, ch
```

in `optimizing_split`.